### ISSUE TITLE
No default CA store for CLI?

### ISSUE THREAD

**Issue by linde12** [denoland/deno#6355]
Hi!

Trying to use deno behind a corporate proxy with im getting `error trying to connect: invalid certificate: UnknownIssuer` - seems like deno isn't locating my default CA store (e.g. like curl does)

I know i can specify a path to a cert, e.g. `deno cache --cert /path/to/ca.crt https://deno.land/...` but this is tedious.

So a couple of questions come to mind:

- Does deno have a built-in CA store (i didn't see it try to open anything after running `strace`) or none at all?
- Is there some way of configuring your cert permanently? I'm thinking of something like `.npmrc` a la deno, although i feel like deno is trying to avoid things like this.
- Will/might deno in the future find the default CA store of the system, or is it a design decision to not do this?
- Are there any other "workarounds" to this problem? I don't want to have to specify the path to my CA file all the time.

Thanks!

**Comment by ry** [comment]
> Is there some way of configuring your cert permanently? I'm thinking of something like .npmrc a la deno, although i feel like deno is trying to avoid things like this.

Yes we avoid having configuration files. I'd be open to having a DENO_CERT environmental variable.

> Will/might deno in the future find the default CA store of the system, or is it a design decision to not do this?

Yes I'm open to it. TBH I'm not sure how we handle default CAs currently - we use [rustls](https://github.com/ctz/rustls) - I assume it must be statically included in the crate if you're not seeing it in strace.

**Comment by linde12** [comment]
@ry Ok, great!

Maybe both could be useful? I'm new to rust & deno, but for the environment variable i thought we could do something like this. What do you think?


```diff
diff --git a/cli/flags.rs b/cli/flags.rs
index 4226bb5c..e5b047df 100644
--- a/cli/flags.rs
+++ b/cli/flags.rs
@@ -181,6 +181,7 @@ static ENV_VARIABLES_HELP: &str = "ENVIRONMENT VARIABLES:
     DENO_DIR             Set the cache directory
     DENO_INSTALL_ROOT    Set deno install's output directory
                          (defaults to $HOME/.deno/bin)
+    DENO_CERT            Load certificate authority from PEM encoded file
     NO_COLOR             Set to disable color
     HTTP_PROXY           Proxy address for HTTP requests
                          (module downloads, fetch)
diff --git a/cli/global_state.rs b/cli/global_state.rs
index 959d794c..a2a2b676 100644
--- a/cli/global_state.rs
+++ b/cli/global_state.rs
@@ -58,6 +58,7 @@ impl GlobalState {
     let dir = deno_dir::DenoDir::new(custom_root)?;
     let deps_cache_location = dir.root.join("deps");
     let http_cache = http_cache::HttpCache::new(&deps_cache_location);
+    let ca_file = flags.ca_file.clone().or_else(|| env::var("DENO_CERT").map(String::into).ok());

     let file_fetcher = SourceFileFetcher::new(
       http_cache,
@@ -65,7 +66,7 @@ impl GlobalState {
       flags.cache_blocklist.clone(),
       flags.no_remote,
       flags.cached_only,
-      flags.ca_file.clone(),
+      ca_file,
     )?;

     let ts_compiler = TsCompiler::new(
```

**Comment by ry** [comment]
@linde12 LGTM on first glance. Would you be able to submit a PR for that?

**Comment by bartlomieju** [comment]
Resolved in #6370
